crates/rpc/rpc/src/eth/api/fees.rs

```diff
@@ -100,110 +101,83 @@ where
         // Treat a request for 1 block as a request for `newest_block..=newest_block`,
         // otherwise `newest_block - 2
         // SAFETY: We ensured that block count is capped
+
         let start_block = end_block_plus - block_count;
-        let headers = self.provider().sealed_headers_range(start_block..=end_block)?;
-        if headers.len() != block_count as usize {
-            return Err(EthApiError::InvalidBlockRange)
+
+        let mut fee_entries = self.fee_history_cache().get_history(start_block, end_block).await?;
+        if fee_entries.is_empty() {
+            for block in self.provider().block_range(start_block..=end_block)?.iter() {
+                let mut entry = FeeHistoryEntry::new(block, block.clone().seal_slow().hash);
+                let percentiles = predefined_percentiles();
+
+                entry.rewards = calculate_reward_percentiles_for_block(
+                    &percentiles,
+                    &entry,
+                    self.cache().clone(),
+                )
+                .await?;
+                fee_entries.push(entry);
+            }
+            // calculate rewards on the fly
         }
 
+        if fee_entries.len() != block_count as usize {
+            return Err(EthApiError::InvalidBlockRange)
+        }
         // Collect base fees, gas usage ratios and (optionally) reward percentile data
         let mut base_fee_per_gas: Vec<U256> = Vec::new();
         let mut gas_used_ratio: Vec<f64> = Vec::new();
         let mut rewards: Vec<Vec<U256>> = Vec::new();
-        for header in &headers {
-            base_fee_per_gas
-                .push(U256::try_from(header.base_fee_per_gas.unwrap_or_default()).unwrap());
-            gas_used_ratio.push(header.gas_used as f64 / header.gas_limit as f64);
 
-            // Percentiles were specified, so we need to collect reward percentile ino
+        for entry in &fee_entries {
+            base_fee_per_gas.push(U256::try_from(entry.base_fee_per_gas).unwrap());
+            gas_used_ratio.push(entry.gas_used_ratio);
+
             if let Some(percentiles) = &reward_percentiles {
-                rewards.push(self.calculate_reward_percentiles(percentiles, header).await?);
+                let mut block_rewards = Vec::new();
+
+                for &percentile in percentiles.iter() {
+                    block_rewards.push(self.fetch_reward_for_percentile(entry, percentile));
```

Comment:
this is only necessary if read them from the cache, if they were out of range we don't have to do this.